lib/db/mysql/index.js

```diff
@@ -189,6 +190,12 @@ const QUERY_CODE_DELETE_USER = 'DELETE FROM codes WHERE userId=?';
 const QUERY_DEVELOPER = 'SELECT * FROM developers WHERE email=?';
 const QUERY_DEVELOPER_DELETE = 'DELETE FROM developers WHERE email=?';
 const QUERY_PURGE_EXPIRED_TOKENS = 'DELETE FROM tokens WHERE clientId NOT IN (?) AND expiresAt < NOW() LIMIT ?;';
+// N.B., this sort order with limit is not actually fully deterministic
+const QUERY_EXPIRED_TOKENS =
+  'SELECT expiresAt, token, clientId FROM tokens WHERE expiresAt > ? AND expiresAt < NOW() ORDER BY expiresAt ASC, clientId, token LIMIT ?';
```

Comment:
Just to elaborate a bit on that cryptic comment, when I run the script against a million row table (generated with randomized expiresAt times), at then end, none of the protected clientId's are deleted and **almost** all of the target clientId rows **are** deleted.

But a few hundred remain.

The target rows that haven't been deleted are in tight bands of expiresAt time that are close to the end of each time window. I've dumped out all the rows that are found by each select and, yep, these "boundary" rows aren't found. But they should have been!. They have expiresAt that are older than rows that were found. I can't explain exactly why except to say 'indeterminate' (and some resources on the web do say that this type of range query with limit may exhibit this behaviour if using a non-unique index, expiresAt).

Anyways, I think this purgeExpiredTokensById method can be used to eliminate the bulk of the backlog and get the table size down to something less huge. And then the original version of purgeExpiredTokens can be used to pick up the stragglers.

Reply:
Or I have a minor bug in my code. But I don't see it. ;-)

Reply:
Not sure it explains your symptons, but I feel like one of these `>` or `<` should be a `>=` or `<=`.

More generally, what happens if the `LIMIT 200` happens to cut off the result set halfway through a set of tokens that all expire at the same time?  E.g. if there are 300 tokens that all expire at the same time, does this delete the first 200 of them and then skip of the rest?

Also, we might want to expand the index to include the `token` and `clientId` columns so the sort can be satisfied completely from the index:

```
mysql> explain SELECT expiresAt, token, clientId FROM tokens WHERE expiresAt > '2018-00-00' AND expiresAt < NOW() ORDER BY expiresAt ASC, clientId, token LIMIT 200;
+----+-------------+--------+------+---------------+------+---------+------+------+-----------------------------+
| id | select_type | table  | type | possible_keys | key  | key_len | ref  | rows | Extra                       |
+----+-------------+--------+------+---------------+------+---------+------+------+-----------------------------+
|  1 | SIMPLE      | tokens | ALL  | idx_expiresAt | NULL | NULL    | NULL |   23 | Using where; Using filesort |
+----+-------------+--------+------+---------------+------+---------+------+------+-----------------------------+
1 row in set, 1 warning (0.01 sec)
```

Reply:
Yeah, could add that index. But if this job can run productively enough without it, I'd rather defer adding an index until I've reduced the row count (or not adding any index, if only this query would benefit. Actually, now that I think of it, the `client, token` part was only added as I flailed around trying to "fix" the ordering that was causing my "bug").